### PR TITLE
Provide Rejection Documentation #593

### PR DESCRIPTION
**Pull request by jlprat**
Issue: #593
* Extend `RejectionHandler` to have `mapRejectionResponse`
* Add tests for it
* Documenation for Rejections adapted to Java APIs
* Fix code example for Scala docs

### CODE REVIEW

File: akka-http/src/main/scala/akka/http/javadsl/server/RejectionHandler.scala [hunk 1]
```diff
@@ -27,13 +31,18 @@ final class RejectionHandler(val asScala: server.RejectionHandler) {
    * "Seals" this handler by attaching a default handler as fallback if necessary.
    */
   def seal = new RejectionHandler(asScala.seal)
+
+  /** Map any HTTP response which was returned by this RejectionHandler to a different one before rendering it. */
+  def mapRejectionResponse(map: function.UnaryOperator[HttpResponse]): RejectionHandler = new RejectionHandler(asScala.mapRejectionResponse(resp ⇒ {
```
Comment (jlprat) [inline akka-http/src/main/scala/akka/http/javadsl/server/RejectionHandler.scala]: Note this method was missing in the Java API

File: akka-http/src/main/scala/akka/http/javadsl/server/RejectionHandler.scala [hunk 2]
```diff
@@ -27,13 +31,18 @@ final class RejectionHandler(val asScala: server.RejectionHandler) {
...
+    JavaMapping.toScala(map.apply(resp))
+  }))
 }
 
 class RejectionHandlerBuilder(asScala: server.RejectionHandler.Builder) {
   def build = new RejectionHandler(asScala.result())
 
   /**
-   * Handles a single [[Rejection]] with the given partial function.
+   * Handles a single [[Rejection]] with the given function.
```
Comment (ktoso) [inline akka-http/src/main/scala/akka/http/javadsl/server/RejectionHandler.scala]: good catch

File: docs/src/main/paradox/java/http/routing-dsl/rejections.md [hunk 3]
```diff
@@ -19,13 +17,13 @@ another route that can complete it. If there are more rejections all of them wil
...
 can be used to convert a set of rejections into an `HttpResponse` (which, in most cases, will be an error response).
-`Route.seal` internally wraps its argument route with the @ref[handleRejections](directives/execution-directives/handleRejections.md#handlerejections-java) directive in order to "catch"
+`Route.seal()` internally wraps its argument route with the @ref[handleRejections](directives/execution-directives/handleRejections.md#handlerejections-java) directive in order to "catch"
 and handle any rejection.
 
 ## Predefined Rejections
 
 A rejection encapsulates a specific reason why a route was not able to handle a request. It is modeled as an object of
-type `Rejection`. Akka HTTP comes with a set of @github[predefined rejections](/akka-http/src/main/scala/akka/http/scaladsl/server/Rejection.scala), which are used by the many
+type `Rejection`. Akka HTTP comes with a set of @github[predefined rejections](/akka-http/src/main/scala/akka/http/javadsl/server/Rejections.scala), which are used by the many
```
Comment (ktoso) [inline docs/src/main/paradox/java/http/routing-dsl/rejections.md]: good catch again :)

File: docs/src/main/paradox/java/http/routing-dsl/rejections.md [hunk 4]
```diff
@@ -86,28 +77,28 @@ So, for example, if the @ref[path](directives/path-directives/path.md#path-java)
...
+@@snip [RejectionHandlerExamplesTest.java](../../../../../test/java/docs/http/javadsl/server/RejectionHandlerExamplesTest.java) { #custom-handler-example-java }
 
-The easiest way to construct a `RejectionHandler` is via the `RejectionHandler.Builder` that Akka HTTP provides.
+The easiest way to construct a `RejectionHandler` is via the `RejectionHandlerBuilder` class that Akka HTTP provides.
 After having created a new `Builder` instance with `RejectionHandler.newBuilder()`
 you can attach handling logic for certain types of rejections through three helper methods:
 
-handle
-: Handles certain rejections with the given partial function. The partial function simply produces a `Route` which is
+handle(Class<T>, Function<T, Route>)
```
Comment (ktoso) [inline docs/src/main/paradox/java/http/routing-dsl/rejections.md]: cool

File: docs/src/main/paradox/java/http/routing-dsl/rejections.md [hunk 5]
```diff
@@ -122,13 +113,26 @@ itself.
...
+@@snip [RejectionHandlerExamplesTest.java](../../../../../test/java/docs/http/javadsl/server/RejectionHandlerExamplesTest.java) { #example-json }
 
 #### Adding the unmatched route in handleNotFound
 
 Since rejection handlers are routes themselves, it is possible to do anything you could possibly want inside such handler.
 For example you may want to include the path which was not found in the response to the client, this is as simple as 
 using the `extractUnmatchedPath` and completing the route with it.
 
-@@snip [RejectionHandlerExamplesSpec.scala](../../../../../test/java/docs/http/javadsl/server/directives/ExecutionDirectivesExamplesTest.java) { #handleNotFoundWithDefails }
+@@snip [ExecutionDirectivesExamplesTest.java](../../../../../test/java/docs/http/javadsl/server/directives/ExecutionDirectivesExamplesTest.java) { #handleNotFoundWithDefails }
```
Comment (ktoso) [inline docs/src/main/paradox/java/http/routing-dsl/rejections.md]: good catch

**Comment by ktoso** [comment]
LGTM looks good, thanks a ton - will be in todays release

**Pull request merged**